src/plans/D3MCompoundDaiPlan.sol

```diff
@@ -0,0 +1,199 @@
+    }
+    function _wmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, y) / WAD;
+    }
+    function _wdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, WAD) / y;
+    }
+
+    // --- Administration ---
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "barb") {
+            barb = data;
+        } else revert("D3MCompoundDaiPlan/file-unrecognized-param");
+        emit File(what, data);
+    }
+    function file(bytes32 what, address data) external auth {
+        if (what == "rateModel") rateModel = InterestRateModelLike(data);
+        else revert("D3MCompoundDaiPlan/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function getTargetAssets(uint256 currentAssets) external override view returns (uint256) {
+        uint256 targetInterestRate = barb;
+        if (targetInterestRate == 0) return 0; // De-activated
+
+        uint256 borrows = cDai.totalBorrows();
+        uint256 totalPoolSize = _sub(_add(cDai.getCash(), borrows), cDai.totalReserves());
+
+        uint256 targetTotalPoolSize = _calculateTargetSupply(targetInterestRate, borrows);
+
+        if (targetTotalPoolSize >= totalPoolSize) {
+            // Increase debt (or same)
+            return _add(currentAssets, targetTotalPoolSize - totalPoolSize);
+        } else {
+            // Decrease debt
+            uint256 decrease = totalPoolSize - targetTotalPoolSize;
+            if (currentAssets >= decrease) {
+                return currentAssets - decrease;
+            } else {
+                return 0;
+            }
+        }
+    }
+
+    // TODO: this function seems unneeded, remove once it's removed from the interface and AAVE
+    // targetSupply = cash + borrows - reserves
+    function calculateTargetSupply(uint256 targetInterestRate) external view returns (uint256) {
+        return _calculateTargetSupply(targetInterestRate, cDai.totalBorrows());
+    }
+
+    function _calculateTargetSupply(uint256 targetInterestRate, uint256 borrows) internal view returns (uint256) {
+        uint256 kink                   = rateModel.kink();
+        uint256 multiplierPerBlock     = rateModel.multiplierPerBlock();
+        uint256 baseRatePerBlock       = rateModel.baseRatePerBlock();
+        uint256 jumpMultiplierPerBlock = rateModel.jumpMultiplierPerBlock();
+
+        uint256 normalRate = _add(_wmul(kink, multiplierPerBlock), baseRatePerBlock);
+
+        uint256 targetUtil;
+        if (targetInterestRate > normalRate) {
+            if (jumpMultiplierPerBlock == 0) return 0; // illegal rate, max is normal rate for this case
+            targetUtil = _add(kink, _wdiv(targetInterestRate - normalRate, jumpMultiplierPerBlock));             // (1)
+        } else if (targetInterestRate > baseRatePerBlock) {
+            targetUtil = _wdiv(targetInterestRate - baseRatePerBlock, multiplierPerBlock);                       // (2)
```

Comment:
Ok calculation makes sense. Still there is a discrepancy with how Aave is handled in that bad values of `bar` will cause this function to revert in Aave vs in here it will just return 0. We should unify this design. I think using your conversion of illegal values to return 0 target supply instead of reverting is a good idea.

Reply:
Resolving this as implemented on AAVE V2 branch.